zcash_client_backend/src/data_api.rs

```diff
@@ -560,6 +560,9 @@ pub trait WalletRead {
         account: AccountId,
         max_height: BlockHeight,
     ) -> Result<HashMap<TransparentAddress, Amount>, Self::Error>;
+
+    /// Returns a vector with the IDs of all accounts known to this wallet.
+    fn get_account_ids(&self) -> Result<Vec<AccountId>, Self::Error>;
```

Comment:
Since we're starting to be a bit forward-looking, I'm wondering now whether we shouldn't come up with a better abstraction than `AccountId`; the `AccountId` itself is really a low-level implementation detail, and it assumes a single seed. One could alternately use a tuple `(SeedFingerprint, AccountId)` but that doesn't necessarily support FVK-only or IVK-only use cases very well because in those cases you probably don't have a seed fingerprint available, and so would potentially have to fake it; realistically, you may also not have the account ID for zip32 derivation, so we probably need a different approach here. 

Perhaps a better approach would be to use the UIVK for the account identifier everywhere we need it?

Reply:
The UIVK is way heavier than we need, and it also makes identifying the corresponding account much more complex (because the UIVK doesn't necessarily contain all the same receiver types as the wallet supports, and then we'd have a whole extra axis to handle when a UIVK contains some receivers that do correspond to an account and other receivers that don't, or a UIVK that contains receivers from two accounts under the same or different seeds).

Reply:
The other possible approach, which I like less, is using a database-internal identifier, a UUID or some such, but we've been moving away from the wallet interfaces leaking internal identifiers.

Reply:
Another, even better, possibility: how about we make `AccountId` an associated type on the `WalletRead` trait? Then, for the sqlite implementation for now, we can continue to use the `zip32::AccountId` type, but in the future we can change that type without breaking code for which that type is existentially quantified.

Reply:
Really good questions. I agree this addition can't stand as-is given the use cases we'd like to add later. But I suspect supporting all that will be a far more sweeping change that is likely to be months off, at least. This addresses a need that we have now and later, as the `AccountId` should be replaceable (across all the API both new and pre-existing) across the board with whatever the new shape takes.
So my vote would be in favor of merging the PR (after addressing the feedback), and I look forward to contributing to the additional use cases soon after.

FWIW, my current idea for the future of the accounts table would be one column for the IVK of each individual pool, so that no question of which receivers a UIVK comes up as a cause for an account lookup failure. The primary key of the table shouldn't be any one of these columns because no one column is guaranteed to be non-null. A multi-key PK would be both heavy and not be friendly to lookups either. 
As my understanding of db schema best practices suggests that PKs should _not_ be user data at all, I'd then add a auto-increment PK, that in the rust crate we can expose as a struct with a private field. Callers can use an API to exchange an IVK (or any other kind of key) for the new PK that represents the account, and keep that around for as long as they need to interact with that account, or lookup repeatedly -- whatever is convenient and performant for them.

Reply:
For this PR, we'll leave account IDs as-is. I've opened https://github.com/zcash/librustzcash/issues/1107 for looking into generalising account IDs.